draft/0002-subdocapi.md

```diff
@@ -0,0 +1,380 @@
+API to return values or indicate status.
+
+[The `Document` _may_ be subclassed into `DocumentFragment` to signal that the
+value contained therein is only _part_ of the document (a fragment).
+
+This subclass contains the same fields as the `Document`, except the `value`
+(or `contents`) is replaced by the `fragment` field. This is to avoid confusion
+regarding whether the value "truly" contains the entire document value, or
+merely a subdoc.
+
+The `Document` object itself is the return value from normal KV operations
+which contain the value, cas, mutation token, etc.
+
+This object is used as an example, some SDKs may wish to either use the
+'value' or 'document' field of the existing `Document` return value, or
+subclass it, replacing value with 'fragment'.
+
+## Path
+
+From the perspective of the SDK, the path is a simple string used to specify
+a location within the document. The path uses N1QL syntax to specify locations
+within a document, thus a valid path may be:
+
+    foo.bar.baz
+
+or
+
+    `foo`.`bar`.`baz`
+
+or
+
+    `foo.some_array[0]`
+
+or
+
+    `foo.some_array[-1]`
+
+As in N1QL, special path characters may be escaped via a backtick (`` ` ``).
+
+Some operations (notably, those taking a path to an array) may accept the
+_empty path_ (i.e. an empty string)
+which refers to the entire document. This is valid if the document is a bare
+JSON array.
+
+Each path component must be valid JSON (without the enclosing quotes). This
+means that things like tabs, quotes, and backslashes must be escaped. To think
+of it another way, the path matching algorithm in the server works via simple
+string comparison, so a JSON key `"bs\\ and \t tabs are here" would be matched
+by a path component of `"bs\\ and \t tabs are here"`.
+
+The SDK MUST verify that a path is non-empty for operations which do not
+support empty paths. All other forms of path verification is done on the server
+(but an empty path for unsupported operations will return a memcached `EINVAL`
+which isn't really helpful).
+
+## Single-path Methods
+
+Methods are named exactly like their top-level counterparts, and should also
+exist at the top level as well.
+
+Error codes are common to most operations.
+
+### `get_in(docid, path)`
+Gets a single path from the document. Returns a `Document`
+(or `DocumentFragment`) containing the path value.
+
+* `docid`: Document ID
+* `path`: The sub-document path
+
+### `exists_in(docid, path)`
+
+Like `get_in`, but does not return the value (only checks if it exists)`. If the
+item does not exist, the error code is suitably set in the result (or whatever
+the appropriate error delivery mechanism is for the SDK).
+
+### `upsert_in(docid, path, value, create_parents, cas, ttl, persist_to, replicate_to)`
+Unconditionally places the specified value in the provided path.
+
+* `docid`: Document ID
+* `path`: the sub-document path
+* `value`: the sub-document value. This should be anything JSON-serializable.
```

Comment:
I suppose this would make sense if your language is using a `Fragment` class or similar, however the `Fragment` is not part of the mandated return types (it is merely a suggestion for those languages which are using compound inputs)